src/Microsoft.AspNet.Mvc.ModelBinding/Binders/ComplexModelDtoModelBinder.cs

```diff
@@ -19,8 +19,10 @@ public sealed class ComplexModelDtoModelBinder : IModelBinder
                 var dto = (ComplexModelDto)bindingContext.Model;
                 foreach (var propertyMetadata in dto.PropertyMetadata)
                 {
-                    var propertyModelName = ModelBindingHelper.CreatePropertyModelName(bindingContext.ModelName,
-                                                                                       propertyMetadata.PropertyName);
+                    var propertyModelName =
+                        ModelBindingHelper.CreatePropertyModelName(
+                            bindingContext.ModelName,
+                            propertyMetadata.BinderModelName ?? propertyMetadata.PropertyName);
```

Comment:
Talked about this, and I think the way it's done here is correct. `[FromHeader]` is really an outlier. It's likely that form and query need to deal with heirarchies, routing not so much, but route-data is likely to show up on a top-level model or property.